File: app/src/main/java/com/ucb/eldroid/ecoconnect/ui/bottomnav/dashboard/crowdfunding/CreateProject.java
package com.ucb.eldroid.ecoconnect.ui.bottomnav.dashboard.crowdfunding;

import android.app.DatePickerDialog;
import android.content.Intent;
import android.database.Cursor;
import android.net.Uri;
import android.os.Bundle;
import android.provider.MediaStore;
import android.view.View;
import android.widget.ArrayAdapter;
import android.widget.DatePicker;
import android.widget.ImageView;
import android.widget.Spinner;
import android.widget.TextView;

import androidx.activity.EdgeToEdge;
import androidx.appcompat.app.AppCompatActivity;
import androidx.core.graphics.Insets;
import androidx.core.view.ViewCompat;
import androidx.core.view.WindowInsetsCompat;

import com.ucb.eldroid.ecoconnect.R;

import java.util.Calendar;

public class CreateProject extends AppCompatActivity {

    private static final int IMAGE_PICK_REQUEST_CODE = 1000; // Arbitrary request code

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_create_project);


        Spinner envDataSpinner = findViewById(R.id.envDataSpinner);

        TextView deadlineDateTextView = findViewById(R.id.deadlineDateTextView);


        ArrayAdapter<CharSequence> adapter = ArrayAdapter.createFromResource(
                this,
                R.array.environmental_data_items,
                android.R.layout.simple_spinner_item
        );
        adapter.setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item);
        envDataSpinner.setAdapter(adapter);

        deadlineDateTextView.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                Calendar calendar = Calendar.getInstance();
                DatePickerDialog datePickerDialog = new DatePickerDialog(
                        CreateProject.this,
                        new DatePickerDialog.OnDateSetListener() {
                            @Override
                            public void onDateSet(DatePicker view, int year, int monthOfYear, int dayOfMonth) {
                                // Set the selected date to the TextView
                                String selectedDate = dayOfMonth + "/" + (monthOfYear + 1) + "/" + year;
                                deadlineDateTextView.setText(selectedDate);
                            }
                        },
                        calendar.get(Calendar.YEAR),
                        calendar.get(Calendar.MONTH),
                        calendar.get(Calendar.DAY_OF_MONTH)
                );
                datePickerDialog.show();
            }
        });
    }
    public void onImageClicked(View view) {
        // Create an intent to open the image picker
        Intent intent = new Intent(Intent.ACTION_PICK, MediaStore.Images.Media.EXTERNAL_CONTENT_URI);
        intent.setType("image/*");  // Filter for images
        startActivityForResult(intent, IMAGE_PICK_REQUEST_CODE); // Start the image picker activity
    }
    @Override
    protected void onActivityResult(int requestCode, int resultCode, Intent data) {
        super.onActivityResult(requestCode, resultCode, data);

        if (requestCode == IMAGE_PICK_REQUEST_CODE && resultCode == RESULT_OK && data != null) {
            // Get the image URI
            Uri selectedImageUri = data.getData();

            // Optionally, you can set the image in an ImageView
            ImageView imageView = findViewById(R.id.addImage); // Your ImageView where you want to display the image
            imageView.setImageURI(selectedImageUri);

            // If you want to get the path of the selected image, you can use:
            String imagePath = getImagePathFromUri(selectedImageUri);
            // Do something with the image path, like uploading the image to a server
        }
    }

    public String getImagePathFromUri(Uri uri) {
        String[] projection = {MediaStore.Images.Media.DATA};
        Cursor cursor = getContentResolver().query(uri, projection, null, null, null);
        if (cursor != null) {
            cursor.moveToFirst();
            int columnIndex = cursor.getColumnIndex(projection[0]);
            String imagePath = cursor.getString(columnIndex);
            cursor.close();
            return imagePath;
        }
        return null;
    }
}

File: app/src/main/java/com/ucb/eldroid/ecoconnect/ui/bottomnav/dashboard/crowdfunding/CreateProject.kt
package com.ucb.eldroid.ecoconnect.ui.bottomnav.dashboard.crowdfunding

import android.app.Activity
import android.app.DatePickerDialog
import android.content.Context
import android.content.Intent
import android.net.Uri
import android.os.Bundle
import android.provider.MediaStore
import android.widget.Button
import android.widget.EditText
import android.widget.ImageView
import android.widget.TextView
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import com.ucb.eldroid.ecoconnect.R
import com.ucb.eldroid.ecoconnect.data.ApiService
import com.ucb.eldroid.ecoconnect.data.models.Project
import com.ucb.eldroid.ecoconnect.ui.auth.Login
import com.ucb.eldroid.ecoconnect.utils.RetrofitClient
import okhttp3.ResponseBody
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response
import java.util.Calendar
import java.util.UUID

class CreateProject : AppCompatActivity() {

    private var imageUri: Uri? = null // To store selected image URI

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_create_project)

        val deadlineDateTextView = findViewById<TextView>(R.id.deadlineDateTextView)

        deadlineDateTextView.setOnClickListener {
            val calendar = Calendar.getInstance()
            val datePickerDialog = DatePickerDialog(
                this@CreateProject,
                { view, year, monthOfYear, dayOfMonth ->
                    val selectedDate = "$dayOfMonth/${monthOfYear + 1}/$year"
                    deadlineDateTextView.text = selectedDate
                },
                calendar[Calendar.YEAR],
                calendar[Calendar.MONTH],
                calendar[Calendar.DAY_OF_MONTH]
            )
            datePickerDialog.show()
        }

        // Create project button click action
        val createProjectButton = findViewById<Button>(R.id.publishButton)
        createProjectButton.setOnClickListener {
            val token = getAuthToken()
            if (token.isNullOrEmpty()) {
                // Prompt the user to log in if token is not available
                Toast.makeText(this, "Please log in to create a project", Toast.LENGTH_SHORT).show()
                startActivity(Intent(this, Login::class.java)) // Navigate to login screen
                finish()
            } else {
                createProject(token)  // Pass token to create project if authenticated
            }
        }

        // Add an image picker button (Assuming you have a button for this)
        val imagePickerButton = findViewById<ImageView>(R.id.addImage)
        imagePickerButton.setOnClickListener {
            pickImage()
        }
    }

    private fun pickImage() {
        val intent = Intent(Intent.ACTION_PICK, MediaStore.Images.Media.EXTERNAL_CONTENT_URI)
        startActivityForResult(intent, IMAGE_PICK_REQUEST_CODE)
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        super.onActivityResult(requestCode, resultCode, data)
        if (resultCode == Activity.RESULT_OK && requestCode == IMAGE_PICK_REQUEST_CODE) {
            data?.data?.let { uri ->
                imageUri = uri
                // You can update your UI here to display the selected image
                // For example, set it to an ImageView
                val imageView = findViewById<ImageView>(R.id.addImage)
                imageView.setImageURI(uri)
            }
        }
    }

    private fun createProject(token: String) {
        val title = findViewById<EditText>(R.id.titleProject).text.toString()
        val description = findViewById<EditText>(R.id.projectDescription).text.toString()
        val goalAmount = findViewById<EditText>(R.id.moneyGoal).text.toString().toDoubleOrNull() ?: 0.0
        val deadline = findViewById<TextView>(R.id.deadlineDateTextView).text.toString()
        val imagePath = imageUri?.let { getImagePathFromUri(it) } // Use the selected image URI

        val project = Project(
            projectId = UUID.randomUUID().toString(), // Random UUID for projectId
            title = title,
            description = description,
            goalAmount = goalAmount,
            deadline = deadline,
            imagePath = imagePath
        )

        // Send the project to the backend
        sendProjectToBackend(project, token)
    }

    private fun sendProjectToBackend(project: Project, token: String) {
        val retrofit = RetrofitClient.instance
        val apiService = retrofit?.create(ApiService::class.java)

        // Pass the token as part of the Authorization header
        val call = apiService?.createProject(project, "Bearer $token")
        call?.enqueue(object : Callback<ResponseBody?> {
            override fun onResponse(call: Call<ResponseBody?>, response: Response<ResponseBody?>) {
                if (response.isSuccessful) {

                    Toast.makeText(this@CreateProject, "Project successfully created!", Toast.LENGTH_SHORT).show()
                } else {
                    Toast.makeText(this@CreateProject, "Failed to create project", Toast.LENGTH_SHORT).show()
                }
            }

            override fun onFailure(call: Call<ResponseBody?>, t: Throwable) {
                Toast.makeText(this@CreateProject, "Error: ${t.message}", Toast.LENGTH_SHORT).show()
            }
        })
    }

    private fun getImagePathFromUri(uri: Uri?): String? {
        // Get image path from URI (if you have image selection functionality implemented)
        return uri?.let {
            val projection = arrayOf(MediaStore.Images.Media.DATA)
            val cursor = contentResolver.query(uri, projection, null, null, null)
            cursor?.use {
                it.moveToFirst()
                val columnIndex = it.getColumnIndex(projection[0])
                return it.getString(columnIndex)
            }
        }
    }

    private fun getAuthToken(): String? {
        val sharedPreferences = getSharedPreferences("AuthPreferences", Context.MODE_PRIVATE)
        return sharedPreferences.getString("AUTH_TOKEN", null) // Retrieve the stored token
    }

    companion object {
        private const val IMAGE_PICK_REQUEST_CODE = 1000
    }
}